### PR TITLE
Regenerate-unaudited?

### PR DESCRIPTION
**Pull request by Gankra**

### CODE REVIEW

File: src/tests.rs [hunk 1]
```diff
@@ -2476,6 +2667,230 @@ fn builtin_simple_unaudited_in_direct_full() {
...
+
+    config.unaudited.insert(
+        "third-party1".to_string(),
+        vec![unaudited(ver(3), SAFE_TO_DEPLOY)],
+    );
+
+    config.unaudited.insert(
+        "transitive-third-party1".to_string(),
+        vec![unaudited(ver(4), SAFE_TO_RUN)],
+    );
```
Comment (Gankra) [inline src/tests.rs:2734]: this fails to vet

**Comment by Gankra** [comment]
I've add dependency_criteria to unaudited entries as we discussed. The algorithm is the same as before and will still try to be "too smart". I'm coming around to your idea of it only ever picking one of 3 options:

* Preserve an unaudited entry
* Remove an unaudited entry
* Add an unaudited entry that == cur_version

However I still have some concerns about cascading rewrites. For instance, say we have a situation like

```
A:1.0.0 safe-to-deploy, partially audited, requires B to safe-to-run
B:1.0.0 safe-to-run, partially audited
```

Someone now updates A to 2.0.0 in the Cargo.lock, and so everything fails to vet. They decide the best solution to this situation is to run `regenerate-unaudited`. It sees that A is disconnected and emits:

```
[unaudited.A] 2.0.0 safe-to-deploy
```

But now B fails to vet, so it also adds:

```
[unuaudited.B] 1.0.0 safe-to-deploy
```

But now B's only-safe-to-run dependencies also fail, so it proceeds to re-emit `unaudited` entries for *everything* in that subtree!

**Comment by Gankra** [comment]
Testing this out on m-c and seeing some weird bugs. Investigating...

**Comment by Gankra** [comment]
There's something messed up going on with the `bitsdownloader` subtree in gecko. We seemingly seem to understand it, but we completely fail to `suggest` or `suggest --guess-deeper` a couple packages that are currently part of it.

As a result, regenerate-unaudited fails to emit *anything* for these packages and a subsequent `vet` fails:

```txt
$ cargo vet
Vetting Failed!

3 unvetted dependencies:
  filetime_win:0.2.0 missing ["safe-to-deploy"]
  guid_win:0.2.0 missing ["safe-to-deploy"]
  replace_with:0.1.7 missing ["safe-to-deploy"]

recommended audits for safe-to-deploy:
    cargo vet inspect guid_win 0.2.0      (used by bits, bits_client)  (481 lines)
    cargo vet inspect filetime_win 0.2.0  (used by bits)               (564 lines)
    cargo vet inspect replace_with 0.1.7  (used by l10nregistry)       (1093 lines)
```

The reverse-deps of guid_win looks like this:

```mermaid
graph LR
  subgraph workspace-members
     subgraph roots
        bitsdownloader
      end
      bits_client
      bits
    
  end
  subgraph third-party
    guid_win
  end
    bitsdownloader ---> bits_client
    bits_client ---> bits
    bits_client ---> guid_win
    bits ---> guid_win
  
```

Which is not terribly interesting! It's all normal-deps too. My guess is that something is happening below one of these nodes that makes it act weird so we don't look at the children properly..? But I'm having trouble debugging/reducing it.

**Comment by Gankra** [comment]
while trying to debug this I fixed *an* issue but it didn't fix this one, sadly.

**Comment by Gankra** [comment]
Ok just reducing workspace-members down to bitsdownload reproduces the issue, and has a much more tame graph:

```mermaid
graph LR
    subgraph roots
        node5{bitsdownload:0.1.0}
    end
    subgraph workspace-members
        node3[/bits:0.2.0/]
        node4[/bits_client:0.2.0/]
        node20[/moz_task:0.1.0/]
        node22[/nserror:0.1.0/]
        node23[/nsstring:0.1.0/]
        node47[/xpcom:0.1.0/]
        node48[/xpcom_macros:0.1.0/]
    end
    subgraph first-party
        node1[async-task:4.0.3]
        node6[cfg-if:0.1.999]
        node21[mozbuild:0.1.0]
    end
    subgraph third-party
        node0(aho-corasick:0.7.18)
        node2(bitflags:1.3.2)
        node7(cfg-if:1.0.0)
        node8(comedy:0.2.0)
        node9(crossbeam-utils:0.8.8)
        node10(cstr:0.2.10)
        node11(encoding_rs:0.8.31)
        node12(filetime_win:0.2.0)
        node13(futures-task:0.3.21)
        node14(getrandom:0.1.16)
        node15(guid_win:0.2.0)
        node16(lazy_static:1.4.0)
        node17(libc:0.2.126)
        node18(log:0.4.14)
        node19(memchr:2.5.0)
        node24(once_cell:1.10.0)
        node25(ppv-lite86:0.2.16)
        node26(proc-macro2:1.0.38)
        node27(quote:1.0.18)
        node28(rand:0.7.3)
        node29(rand_chacha:0.2.2)
        node30(rand_core:0.5.1)
        node31(rand_hc:0.2.0)
        node32(redox_syscall:0.1.57)
        node33(regex:1.5.5)
        node34(regex-syntax:0.6.25)
        node35(remove_dir_all:0.5.3)
        node36(syn:1.0.94)
        node37(tempfile:3.1.0)
        node38(thin-vec:0.2.5)
        node39(thiserror:1.0.31)
        node40(thiserror-impl:1.0.31)
        node41(threadbound:0.1.3)
        node42(unicode-xid:0.2.3)
        node43(wasi:0.9.0+wasi-snapshot-preview1)
        node44(winapi:0.3.9)
        node45(winapi-i686-pc-windows-gnu:0.4.0)
        node46(winapi-x86_64-pc-windows-gnu:0.4.0)
    end
    node0 --> node19
    node3 --> node8
    node3 --> node12
    node3 --> node15
    node3 --> node44
    node4 --> node3
    node4 --> node8
    node4 --> node15
    node4 --> node16
    node4 --> node28
    node4 --> node33
    node4 --> node37
    node4 --> node39
    node5 --> node4
    node5 --> node8
    node5 --> node9
    node5 --> node17
    node5 --> node18
    node5 --> node20
    node5 --> node22
    node5 --> node23
    node5 --> node47
    node6 --> node7
    node8 --> node44
    node9 --> node7
    node9 --> node16
    node10 --> node26
    node10 --> node27
    node11 --> node7
    node12 --> node8
    node12 --> node44
    node14 --> node7
    node14 --> node17
    node14 --> node43
    node15 --> node8
    node15 --> node44
    node18 --> node7
    node20 --> node1
    node20 --> node10
    node20 --> node13
    node20 --> node17
    node20 --> node18
    node20 --> node22
    node20 --> node23
    node20 --> node39
    node20 --> node47
    node21 --> node24
    node22 --> node21
    node22 --> node23
    node23 --> node2
    node23 --> node11
    node26 --> node42
    node27 --> node26
    node28 --> node14
    node28 --> node17
    node28 --> node29
    node28 --> node30
    node28 --> node31
    node29 --> node25
    node29 --> node30
    node30 --> node14
    node31 --> node30
    node33 --> node0
    node33 --> node19
    node33 --> node34
    node35 --> node44
    node36 --> node26
    node36 --> node27
    node36 --> node42
    node37 --> node6
    node37 --> node17
    node37 --> node28
    node37 --> node32
    node37 --> node35
    node37 --> node44
    node39 --> node40
    node40 --> node26
    node40 --> node27
    node40 --> node36
    node44 --> node45
    node44 --> node46
    node47 --> node10
    node47 --> node17
    node47 --> node21
    node47 --> node22
    node47 --> node23
    node47 --> node38
    node47 --> node41
    node47 --> node48
    node48 --> node16
    node48 --> node21
    node48 --> node26
    node48 --> node27
    node48 --> node36
```

**Comment by Gankra** [comment]
hand-crafting a small subset doesn't help. I'm gonna build out some system that lets us export a graph and reimport it into the test harness for reproducing/reduction.

**Comment by bholley** [comment]
Going to merge this as-is to play with it.

**Pull request merged**